src/middlewared/middlewared/plugins/nvmet/kernel.py

```diff
@@ -0,0 +1,790 @@
+            return None
+
+        return str(index)
+
+    @classmethod
+    def dst_name(cls, entry):
+        return f'{entry[cls.dst_query_keys[0]][cls.dst_query_keys[1]]}'
+
+    @classmethod
+    def create_links(cls, entry: dict):
+        return entry['port']['nvmet_port_enabled']
+
+
+def write_config(config):
+    if not pathlib.Path(NVMET_KERNEL_CONFIG_DIR).exists():
+        return
+
+    # Render operations are context managers that do
+    # 1. Create-style operations
+    # 2. yield
+    # 3. Delete-style operations
+    #
+    # Therefore we can nest them to enfore the necessary
+    # order of operations.
+    with NvmetSubsysConfig.render(config):
+        with NvmetHostConfig.render(config):
+            with NvmetPortConfig.render(config):
+                with NvmetPortReferralConfig.render(config):
+                    with NvmetPortAnaReferralConfig.render(config):
+                        with NvmetHostSubsysConfig.render(config):
+                            with NvmetPortSubsysConfig.render(config):
+                                with NvmetNamespaceConfig.render(config):
+                                    pass
+
+
+def _map_device_path(device_type, device_path):
+    match device_type:
+        case 'FILE':
+            if device_path.startswith('/mnt/'):
+                return device_path
+        case 'ZVOL':
+            if device_path.startswith('zvol/'):
+                return zvol_name_to_path(device_path[5:])
+
+
+def _set_namespace_field(subnqn, nsnum, field, value):
+    p = pathlib.Path(NVMET_KERNEL_CONFIG_DIR,
+                     'subsystems',
+                     subnqn,
+                     'namespaces',
+                     str(nsnum),
+                     field)
+    try:
+        p.write_text(f'{value}\n')
+    except FileNotFoundError:
+        pass
+
+
+def _set_namespace_enable(subnqn, nsnum, value):
+    _set_namespace_field(subnqn, nsnum, 'enable', value)
+
+
+def lock_namespace(data):
+    _set_namespace_enable(data['subsys']['nvmet_subsys_subnqn'], data['nsid'], 0)
+
+
+def unlock_namespace(data):
+    _set_namespace_field(data['subsys']['nvmet_subsys_subnqn'],
+                         data['nsid'],
+                         'device_path',
+                         _map_device_path(data['device_type'], data['device_path']))
+    _set_namespace_enable(data['subsys']['nvmet_subsys_subnqn'], data['nsid'], 1)
+
+
+def resize_namespace(data):
+    _set_namespace_field(data['subsys']['nvmet_subsys_subnqn'],
+                         data['nsid'],
+                         'revalidate_size',
+                         1)
+
+
+def clear_config():
+    config_root = pathlib.Path(NVMET_KERNEL_CONFIG_DIR)
+    if not config_root.exists():
+        return
+
+    for port in pathlib.Path(config_root, 'ports').iterdir():
+        for subsys in pathlib.Path(port, 'subsystems').iterdir():
+            subsys.unlink()
```

Comment:
Why so many unlink() operations and not simply shutil.rmtree?

Reply:
Since this is a kernel `configfs` interface, and not just a regular filesystem, certain operations are not permitted.